pallets/foreign-investments/src/impls.rs

```diff
@@ -218,55 +218,49 @@ impl<T: Config> ForeignInvestment<T::AccountId> for Pallet<T> {
 			true
 		} else {
 			T::PoolInspect::currency_for(investment_id.of_pool())
-				.map(|pool_currency| T::TokenSwaps::valid_pair(currency, pool_currency))
+				.map(|pool_currency| T::Swaps::valid_pair(currency, pool_currency))
 				.unwrap_or(false)
 		}
 	}
 }
 
-pub struct FulfilledSwapOrderHook<T>(PhantomData<T>);
-impl<T: Config> StatusNotificationHook for FulfilledSwapOrderHook<T> {
+pub struct FulfilledSwapHook<T>(PhantomData<T>);
+impl<T: Config> StatusNotificationHook for FulfilledSwapHook<T> {
 	type Error = DispatchError;
-	type Id = T::SwapId;
-	type Status = SwapStateOf<T>;
-
-	fn notify_status_change(swap_id: T::SwapId, swap_state: SwapStateOf<T>) -> DispatchResult {
-		match Swaps::<T>::foreign_id_from(swap_id) {
-			Ok((who, investment_id, action)) => {
-				let pool_currency = pool_currency_of::<T>(investment_id)?;
-				let swapped_amount_in = swap_state.swapped_in;
-				let swapped_amount_out = swap_state.swapped_out;
-				let pending_amount = swap_state.remaining.amount_out;
-
-				if pending_amount.is_zero() {
-					Swaps::<T>::update_id(&who, investment_id, action, None)?;
-				}
+	type Id = (T::AccountId, SwapId<T>);
```

Comment:
What prevents us from using Currency already? Is there more work to be done besides extending the key?

Reply:
Yes, i.e. right now, we can not know in which currency we should collect when a redemption collection when it's notified. `pallet-investment` just notify with the `investment_id`, and having more than one foreign currency we do not know in which currency we should transform that. Even worst, we can collect an amount that comes from a mix of redemptions in different foreign currencies.

We should specify how to proceed in this case. That would imply a new storage with all currencies associated with an `investment_id`(maybe a `Vec` or a `Set`), and iterate them in some order, splitting the pool collected amount into different foreign currencies. Not trivial 😓